types/relay-runtime/lib/store/readInlineData.d.ts

```diff
@@ -0,0 +1,19 @@
+import { GraphQLTaggedNode } from '../query/RelayModernGraphQLTag';
+
+interface _RefType<T> {
+    ' $refType': T;
+}
+interface _FragmentRefs<T> {
+    ' $fragmentRefs': T;
+}
+
+type FragmentRef<T> = T extends _RefType<infer U> ? _FragmentRefs<U> : never;
+
+declare function readInlineData<T extends _RefType<any>>(fragment: GraphQLTaggedNode, ref: FragmentRef<T>): T;
+declare function readInlineData<T extends _RefType<any>>(
+    fragment: GraphQLTaggedNode,
+    ref: FragmentRef<T> | null | undefined,
+): T | null | undefined;
+declare function readInlineData(fragment: GraphQLTaggedNode, ref: null | undefined): null | undefined;
```

Comment:
is this overloading overkill? see https://github.com/facebook/relay/blob/71d2a70839920b7d4f49ecc6e290d2ad03fd08bb/packages/relay-runtime/store/readInlineData.js#L46-L48

i.e. `readInlineData` returns null-y iff it's called with something nully.

though arguably calling it with something statically known to be null-y is silly. what's the right way to express this sort of function that passes through nulls?

Reply:
or do i just remove the overload that takes `FragmentRef<T> | null | undefined`?

Reply:
honestly, I'm trying to figure it out what `readInlineData` does

Reply:
see https://relay.dev/docs/en/graphql-in-relay.html#inline

maybe i should must match the overloads in flow

Reply:
I dunno, it seems like the function isn’t useful in any way when passing in `null`, right? There’s runtime code to guard against it, but that doesn’t mean imo that we need to type it up as valid use-cases. Let's just type what is actually useful?

Reply:
so what _is_ useful is that you can pass in a nullable fragment ref and then get back a nullable piece of data

what's the canonical way to represent that? the point i'm looking to convey is something like:

- if you pass in something non-nullable, you'll get something non-null back
- if you pass in something nullable, you'll get back a nullable fragment

i guess it's probably not worth guarding against someone passing in something statically known to be null? i'm guessing that the typing here is just wrong: https://github.com/facebook/relay/blob/71d2a70839920b7d4f49ecc6e290d2ad03fd08bb/packages/relay-runtime/store/readInlineData.js#L39

Reply:
i took out the null-only override

Reply:
> so what is useful is that you can pass in a nullable fragment ref and then get back a nullable piece of data

Gotcha. The overload seems fine then, or is it not working as expected? Alternatively you could use conditional types to more accurately represent the permutations.

* https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html
* https://artsy.github.io/blog/2018/11/21/conditional-types-in-typescript/

Reply:
nope, this is fine – i'm going to err to the side of following the flow types here

Reply:
That makes sense to me 👍